airflow/executors/dask_executor.py

```diff
@@ -81,7 +81,22 @@ def airflow_run():
         if not self.client:
             raise AirflowException(NOT_STARTED_MESSAGE)
 
-        future = self.client.submit(airflow_run, pure=False)
+        resources = None
+        if queue is not None:
+            avail_resources = self.client.run_on_scheduler(
+                lambda dask_scheduler: dask_scheduler.resources
+            )
```

Comment:
This technically works and it is there to stay but we usually use the `run_on_scheduler` method only for debugging purposes.
There is also the possibility to disable this method entirely on the scheduler to prohibit execution of malicious code since this would otherwise allow the execution of generic code.

I see two possibilities:

1. Contribute a `get_ressources` method to `distributed` itself. I think this is a fair feature request
2. Use a scheduler plugin (see https://distributed.dask.org/en/latest/plugins.html#distributed.diagnostics.plugin.SchedulerPlugin)

the scheduler plugin and contribution would use the same code it's just about where the code lives. Should be as simple as

```python
class GetResourcesPlugin(SchedulerPlugin):
    def __init__(self, scheduler):
        self.scheduler = scheduler
        self.scheduler.handlers["get-resources"] = self.get_resources

    def get_resources(self):
        return self.scheduler.resources
```

on client side it's then simply

```python
client.scheduler.get_resources
```

It's a bit boilerplate but this ensures that it will always work, even in locked down clusters.

Reply:
Cool, didn't know about the scheduler plugins, so thanks for that. Not sure what I'm doing wrong, but doesn't quite seem to work though,

```
# file.py
class GetResourcesPlugin(SchedulerPlugin):
    def __init__(self, scheduler):
        self.scheduler = scheduler
        self.scheduler.handlers["get-resources"] = self.get_resources

    def get_resources(self):
        return self.scheduler.resources

@click.command()
def dask_setup(scheduler):
    plugin = GetResourcesPlugin(scheduler)
    scheduler.add_plugin(plugin)
    print(scheduler.handlers)  # I see the desired handler added here
    # print(scheduler.get_resources())    # get error since this method does exist, so tried without it too
```

```
# start scheduler
$ dask-scheduler --preload file.py
```

```
c = Client('<scheduler address>')
c.scheduler.get_resources()  # doesn't seem to work
```

Any ideas what I'm doing wrong here?

In any case, I think suggestion #1 makes the most sense though, since knowing what worker resources you have available is relevant to how you would submit tasks from the client, I think it would be a good idea to be able to access this directly.

Reply:
> scheduler.get_resources()

The scheduler will not have the method but the plugin has. By registering a handler you are registering an RPC with the scheduler. 
The `client.scheduler` is, in fact, not the scheduler but a [PooledRPCCall](https://github.com/dask/distributed/blob/9975baa54b5d2079891e93f57ed8e0f402ef89d8/distributed/core.py#L840) which means that any attributes on that object are translated to an RPC to the scheduler. Therefore `client.scheduler.handler_name(handler_argument)` will execute the method registered under handler `handler_name` with argument `handler_argumnet`. In this specific case it should read

```
client.scheduler.get_ressources() # There are no arguments here 
```

In fact, you might need to wrap this with the sync method since technically the RPC call will be async. Try

```python
client.sync(client.scheduler.get_resources())
```

Reply:
Cool thanks! That helped me get it working with a couple extra tweaks.

The `get_resources` method of the plugin class needed to a argument (looks like a comm.tcp.TCP objet being passed in to it). Also needed to call client.sync like so: `client.sync(client.scheduler.get_resources)`, without the invocation of get_resources.

That said, how do you feel about client.get_worker_resources for the reason I listed above? I think that would be the better approach, since I think that's something users should easily be able to lookup.

Reply:
> That said, how do you feel about client.get_worker_resources for the reason I listed above? I think that would be the better approach, since I think that's something users should easily be able to lookup.

As I said, I think this is a sane addition and the code is identical. Since you figured out how this works now, would you be interested in contributing this to distributed?

Reply:
Yea, will open a PR soon. Thanks.

Reply:
@fjetter - PR (with tests) added
https://github.com/dask/distributed/pull/5105